[FLINK-12555] Introduce an encapsulated metric group layout for shuffle API

azagrebin opened the pull request:
## What is the purpose of the change

At the moment, partition/gate create methods in NetworkEnvironment have a lot of metrics arguments to maintain original layout for metric groups. This approach is not quite encapsulated and clean for shuffle API. We can have just one parent group for shuffle metrics. The old layout can be still maintained in parallel and deprecated. At the moment we can do it with a couple of casts (if shuffle implementation is NetworkEnvironment) and adding an additional legacy metric registration which can be removed later.

## Brief change log

  - Change `NetworkEnvironment.createResultPartitionWriters/createInputGates` to have only one parent metric group argument.
  - Add InputChannelMetricsWithLegacy to increment input metrics from the legacy group as well
  - Move legacy metric group creation to deprecated `NetworkEnvironment.registerLegacyNetworkMetrics`
  - clean `Task` of legacy code but add call to `NetworkEnvironment.registerLegacyNetworkMetrics` after creation of partitions and gates (later needs instanceOf check for actual ShuffleService)

## Verifying this change

simple refactoring, existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

azagrebin commented:
The new metric group layout is still to discuss.
We also need to agree on how we deprecate the old groups in docs.

azagrebin commented:
@flinkbot attention @zentol @zhijiangW

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -388,10 +378,12 @@ public Task(
 			metrics.getIOMetricGroup().getNumBytesInCounter());
 
+		// we will have to check type of shuffle service later whether it is NetworkEnvironment
+		//noinspection deprecation
```
inconsistent with other parts of the PR that use `@SuppressWarnings("deprecation")`

azagrebin replied:
it does not work for statements, at least in Idea. that is why Idea has this special extra comment, I guess.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/InputChannelMetrics.java:
```diff
@@ -32,44 +32,79 @@
+
+		@Override
+		public void incNumBytesInLocalCounter(long inc) {
+			super.incNumBytesInLocalCounter(inc);
```
it's quite easy to encapsulate this logic in a `Counter` implementation, which would require significantly less overall changes and would no longer introduce a new special case on how metrics are being used.

```
public static class CounterWrapper implements Counter {
	private final Counter counter1;
	private final Counter counter2;

	private CounterWrapper(Counter counter1, Counter counter2) {
		this.counter1 = counter1;
		this.counter2 = counter2;
	}

	@Override
	public void inc() {
		counter1.inc();
		counter2.inc();
	}

	@Override
	public void inc(long n) {
		counter1.inc(n);
		counter2.inc(n);

	}

	@Override
	public void dec() {
		counter1.dec();
		counter2.dec();
	}

	@Override
	public void dec(long n) {
		counter1.dec(n);
		counter2.dec(n);
	}

	@Override
	public long getCount() {
		// assume that the counters are not accessed directly elsewhere
		return counter1.getCount();
	}
}
```

azagrebin replied:
true, switched to the counter wrapper

azagrebin commented:
Thanks @zentol ! addressed comments

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/InputChannelMetrics.java:
```diff
@@ -72,4 +72,47 @@ public Counter getNumBuffersInLocalCounter() {
+
+		@Override
+		public long getCount() {
+			// assume that the counters are not accessed directly elsewhere
```
I am not sure whether `getCounter` would be actually used. If used we should keep the return counter as previous structure. That means calling `new InputChannelMetrics(parentGroup, networkGroup) instead?

zentol commented:
> And if we think there are no concerns for creating one metric group twice

There are concerns; it prints a warning to the user as it is not how the API is supposed to be used. If any components need access to a shared group then this group should be created once and passed around as needed.
Things were implemented this way on purpose to prevent components from interfering with each other by accident.

zhijiangW commented:
Thanks for the replies @zentol .

But in this PR `parentGroup.addGroup("Network")` is called twice in `NetworkEnvironment#createResultPartitionWriters/InputGates`. So we should change to create it still in task class and then pass it into `createResultPartitionWriters/InputGates` separately?

azagrebin commented:
True, thanks for noticing it @zhijiangW, it was not an intent. I will change it to create group only once. It can be called e.g. `NetworkInput`/`NetworkOutput`.

azagrebin commented:
@zentol
Here are the moved metrics:

|      Deprecated||||           New ||||
|:-|:-|:-|:-|:-|:-|:-:|:-|
|**Type**|**Scope**|**Infix**|**Name**|**Type**|**Scope**|**Infix**|**Name**|
|IO|Task|-|numBytesInLocal|NettyShuffle|Task|NettyShuffleInput|numBytesInLocal|
|IO|Task|-|numBytesInLocalPerSecond|NettyShuffle|Task|NettyShuffleInput|numBytesInLocalPerSecond|
|IO|Task|-|numBytesInRemote|NettyShuffle|Task|NettyShuffleInput|numBytesInRemote|
|IO|Task|-|numBytesInRemotePerSecond|NettyShuffle|Task|NettyShuffleInput|numBytesInRemotePerSecond|
|IO|Task|-|numBuffersInLocal|NettyShuffle|Task|NettyShuffleInput|numBuffersInLocal|
|IO|Task|-|numBuffersInLocalPerSecond|NettyShuffle|Task|NettyShuffleInput|numBuffersInLocalPerSecond|
|IO|Task|-|numBuffersInRemote|NettyShuffle|Task|NettyShuffleInput|numBuffersInRemote|
|IO|Task|-|numBuffersInRemotePerSecond|NettyShuffle|Task|NettyShuffleInput|numBuffersInRemotePerSecond|
|Network|Task|Network.<Input\|Output>.<gate\|partition>|totalQueueLen|NettyShuffle|Task|<NettyShuffleInput\|NettyShuffleOutput>.<gate\|partition>| totalQueueLen |
|Network|Task|Network.<Input\|Output>.<gate\|partition>| minQueueLen |NettyShuffle|Task|<NettyShuffleInput\|NettyShuffleOutput>.<gate\|partition>| minQueueLen |
|Network|Task|Network.<Input\|Output>.<gate\|partition>| maxQueueLen |NettyShuffle|Task|<NettyShuffleInput\|NettyShuffleOutput>.<gate\|partition>| maxQueueLen |
|Network|Task|Network.<Input\|Output>.<gate\|partition>| avgQueueLen |NettyShuffle|Task|<NettyShuffleInput\|NettyShuffleOutput>.<gate\|partition>| avgQueueLen |
|Network|Task| buffers | inputQueueLength |NettyShuffle|Task|NettyShuffleInput.buffers| inputQueueLength |
|Network|Task| buffers | outputQueueLength |NettyShuffle|Task|NettyShuffleOutput.buffers| outputQueueLength |
|Network|Task| buffers | inPoolUsage |NettyShuffle|Task|NettyShuffleInput.buffers| inPoolUsage |
|Network|Task| buffers | outPoolUsage |NettyShuffle|Task|NettyShuffleOutput.buffers| outPoolUsage |

zentol commented:
I'd prefer if the infix were `Shuffle.Netty.Input`/`Shuffle.Netty.Output`, so far we've followed a hierarchical style similarly to out config options.

The PR also needs a rebase.

zentol commented on docs/monitoring/metrics.md:
```diff
@@ -1236,42 +1351,42 @@ Certain RocksDB native metrics are available but disabled by default, you can fi
       <th rowspan="12"><strong>Task</strong></th>
       <td>numBytesInLocal</td>
-      <td>The total number of bytes this task has read from a local source.</td>
+      <td>The total number of bytes this task has read from a local source. Deprecated: see NettyShuffle.</td>
```
let's remove the description entirely and put a nice `attention` icon before the deprecation notice.

zentol commented on docs/monitoring/metrics.md:
```diff
@@ -1071,6 +1072,120 @@ Thus, in order to infer the metric identifier:
   </tbody>
 </table>
 
+### NettyShuffle
```
do users actually know what NettyShuffle is?

azagrebin replied:
True users do not know about `NettyShuffle`, we will still have only one  at the moment.
It could be `Default network shuffle` then.

azagrebin commented:
@zentol  
about infix, we want to avoid multiple metric group creations. Creating outer groups `Shuffle` and `Netty` firstly for partitions then for gates would duplicate it or we have to have execution id indexing of created metrics so far which I would try to avoid.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -359,21 +359,14 @@ public Task(
-		final MetricGroup networkGroup = metrics.getIOMetricGroup().addGroup("Network");
-		final MetricGroup outputGroup = networkGroup.addGroup("Output");
-		final MetricGroup inputGroup = networkGroup.addGroup("Input");
+		MetricGroup taskShuffleMetricGroup = networkEnvironment.createTaskMetricGroup(metrics.getIOMetricGroup());
```
Does it mean `ShuffleEnvironment` interface need provide the method `createTaskMetricGroup`?

azagrebin replied:
My idea was to expose `TaskShuffleContext ShuffleEnvironment.createTaskShuffleContext(taskName, execId, parentMetricGroup)` in `ShuffleEnvironment `. `TaskShuffleContext` would contain  taskName, execId and created in/output metric group as well. `createPartition/Gate` methods would take the `TaskShuffleContext`. This way `ShuffleEnvironment ` can create its own metric group once and other task common parameters are encapsulated.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -228,7 +236,8 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 
-			registerOutputMetrics(outputGroup, buffersGroup, resultPartitions);
+			MetricGroup outputMetricGroup = taskMetricGroup.addGroup(METRIC_GROUP_OUTPUT);
+			registerOutputMetrics(outputMetricGroup, outputMetricGroup.addGroup(METRIC_GROUP_BUFFERS), resultPartitions);
```
Maybe we could only pass `outputMetricGroup` here and make `outputMetricGroup.addGroup(METRIC_GROUP_BUFFERS)` inside `registerOutputMetrics`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -259,11 +269,15 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 			}
 
-			registerInputMetrics(inputGroup, buffersGroup, inputGates);
+			registerInputMetrics(networkInputGroup, networkInputGroup.addGroup(METRIC_GROUP_BUFFERS), inputGates);
```
ditto: create buffers group inside `registerInputMetrics`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -280,6 +294,31 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
+		// similar to MetricUtils.instantiateNetworkMetrics() but inside this IOMetricGroup
+		final MetricGroup networkGroup = metricGroup.addGroup("Network");
+		final MetricGroup outputGroup = networkGroup.addGroup("Output");
+		final MetricGroup inputGroup = networkGroup.addGroup("Input");
```
Use `METRIC_GROUP_OUTPUT` and `METRIC_GROUP_INPUT` instead of `Output` and `Input`, also might define a class var for `Network`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -69,7 +70,13 @@
 
-	private static final String METRIC_GROUP_NETWORK = "Network";
+	@SuppressWarnings("DeprecatedIsStillUsed")
+	@Deprecated private static final String METRIC_GROUP_NETWORK_DEPRECATED = "Network";
```
Put `@Deprecated` in separate line?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -69,7 +70,13 @@
+	private static final String METRIC_GROUP_NETTY = "Netty";
+	private static final String METRIC_GROUP_OUTPUT = "Output";
+	private static final String METRIC_GROUP_INPUT = "Input";
+	private static final String METRIC_GROUP_BUFFERS = "Buffers";
```
There are already too many metrics here and it seems hard to distinguish them if not focusing on the below implementations. Could we give some comments or other forms to clearly know the architecture of these metrics? 
At least to distinguish the task-level metrics and global metrics.

zhijiangW replied:
It is more clearly now 👍

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -280,6 +294,31 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
+		// add metrics for buffers
+		final MetricGroup buffersGroup = metricGroup.addGroup("buffers");
+
+		// similar to MetricUtils.instantiateNetworkMetrics() but inside this IOMetricGroup
```
It seems we could not see `IOMetricGroup` here, might need adjust the annotation.

azagrebin replied:
the comment is still relevant but I will put `metricGroup` in parenthesis:
`.. inside this IOMetricGroup (metricGroup)`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -280,6 +294,31 @@ private void registerInputMetrics(MetricGroup inputGroup, MetricGroup buffersGro
+	public void registerLegacyNetworkMetrics(
+		MetricGroup metricGroup,
+		ResultPartitionWriter[] producedPartitions,
+		InputGate[] inputGates) {
```
One more indentation for parameters to keep inconsistent with other methods

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -160,10 +169,10 @@ public static NetworkEnvironment create(
 	}
 
-	private static void registerNetworkMetrics(MetricGroup metricGroup, NetworkBufferPool networkBufferPool) {
+	private static void registerNetworkMetrics(String groupName, MetricGroup metricGroup, NetworkBufferPool networkBufferPool) {
```
It might be better to put this method together with other metric related methods below, avoid dispersed anywhere.

azagrebin replied:
I would avoid history change in this commit, could be a followup to put all metric methods into a `NettyShuffleMetricFactory` with constants.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -388,16 +381,18 @@ public Task(
 
+		// we will have to check type of shuffle service later whether it is NetworkEnvironment
+		//noinspection deprecation
+		networkEnvironment.registerLegacyNetworkMetrics(metrics.getIOMetricGroup(), resultPartitionWriters, gates);
```
ditto: it seems a bit trouble after task only sees `ShuffleEnvironment` future, otherwise we need expose this method in interface.

zentol replied:
Since we intend to remove this in the next release It should be fine to use a cast here, so that we don't have to add such a short-lived method to the interface.

azagrebin replied:
which is now happening after rebasing onto #8608

zhijiangW commented:
@azagrebin thanks for the updates and I left some other small comments.

azagrebin commented:
Thanks for the review @zhijiangW, I have addressed comments

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java:
```diff
@@ -70,18 +70,35 @@
 	private static final String METRIC_OUTPUT_QUEUE_LENGTH = "outputQueueLength";
 	private static final String METRIC_OUTPUT_POOL_USAGE = "outPoolUsage";
+
+	// task level output metrics: Shuffle.Netty.Input.*
```
output -> input

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -0,0 +1,148 @@
+	public static void registerShuffleMetrics(
+			MetricGroup metricGroup,
+			NetworkBufferPool networkBufferPool) {
+		checkNotNull(metricGroup);
```
checkNotNull(networkBufferPool)?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/metrics/NettyShuffleMetricFactory.java:
```diff
@@ -0,0 +1,148 @@
+	private static final String METRIC_INPUT_QUEUE_LENGTH = "inputQueueLength";
+	private static final String METRIC_INPUT_POOL_USAGE = "inPoolUsage";
+
+	private NettyShuffleMetricFactory() {
```
This constructor seems not necessary

azagrebin replied:
This is an indicator that it is a utility stateless class with only static methods which is not supposed to be instantiated.

zhijiangW commented:
Thanks for the updates @azagrebin ! 
I like the way of metric factory to make `NetworkEnvironment` simple. Only several nit comments.

azagrebin commented:
Thanks for the reviews @zentol @zhijiangW, I've addressed comments
This PR is now based on #8608 which I think needs to be merged first.
While rebasing, I introduced a `ShuffleIOOwnerContext` which is now created by Task using `ShuffleEnvironment.createShuffleIOOwnerContext` before creating partitions/gates. `NettyShuffleEnvironment.createShuffleIOOwnerContext` also creates netty metric groups, only once.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -359,21 +360,13 @@ public Task(
-		final MetricGroup outputGroup = networkGroup.addGroup("Output");
-		final MetricGroup inputGroup = networkGroup.addGroup("Input");
+		ShuffleIOOwnerContext taskShuffleContext = shuffleEnvironment
+			.createShuffleIOOwnerContext(taskNameWithSubtaskAndId, executionId, metrics.getIOMetricGroup());
```
The formatting:
`shuffleEnvironment.createShuffleIOOwnerContext(
			taskNameWithSubtaskAndId,
			executionId,
			metrics.getIOMetricGroup());
`

azagrebin replied:
In this case we do not break the arg list, the whole function is on the new line, same as chained calls.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -359,21 +360,13 @@ public Task(
+			.createShuffleIOOwnerContext(taskNameWithSubtaskAndId, executionId, metrics.getIOMetricGroup());
 
 		// produced intermediate result partitions
 		final ResultPartitionWriter[] resultPartitionWriters = shuffleEnvironment.createResultPartitionWriters(
```
remove final for `resultPartitionWriters` for consistent with other temp vars?

azagrebin replied:
this file actually looks like more using finals atm.

zhijiangW replied:
But the vars of `taskShuffleContext` and following `gates` are not defined as final. ATM only `taskNameWithSubtaskAndId` and `resultPartitionWriters` are final. Or we could make the new added `taskShuffleContext` as final to keep consistency.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -359,21 +360,13 @@ public Task(
 
 		final String taskNameWithSubtaskAndId = taskNameWithSubtask + " (" + executionId + ')';
```
Might remove this local var to use `taskNameWithSubtask + " (" + executionId + ')'` directly below.

azagrebin replied:
I would keep it to not mix concerns in one line.

zhijiangW replied:
I could agree your way to keep it. :)

My previous concern is that `taskNameWithSubtaskAndId` was used in many places before, so it worths defining a local var. Now it is only used in one place and we could reduce one line to make task constructor short to remove it.
Another small consideration is avoiding final issue as above mentioned.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java:
```diff
@@ -92,25 +92,34 @@
+	 * <p>This method has to be called only once to avoid duplicated internal metric group registration.
+	 *
+	 * @param ownerName the owner name, used for logs
+	 * @param executionAttemptID execution attempt id of the producer or consumer
```
of the producer or consumer -> of the owner?

azagrebin replied:
At the moment we have exec id only for tasks/executions, in this context producer or consumer. I thought about owner as a potentially any shuffle IO component within task/execution, possibly even source/sink

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+			checkNotNull(ownerName),
+			checkNotNull(executionAttemptID),
+			parentGroup,
+			nettyGroup.addGroup(METRIC_GROUP_INPUT),
```
It might be better to put these two `addGroup` covered by above `createShuffleIOOwnerMetricGroup` then return tuple2 to reference here because all the metrics related operations should be done inside `NettyShuffleMetricFactory`

azagrebin replied:
I thought about it, I think the operation is too primitive atm for an extra utility method. Besides, I think that in general `tuple2` is very adhoc and not so readable.

zhijiangW replied:
Actually I also do not like the way of tuple. 
But we provide the method of `createShuffleIOOwnerMetricGroup` in factory, and it is only used for adding group simple. So it might bring confused that which `addGroup` should be done inside factory and which should be done outside. The scope/rule seems not clear.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+			registerOutputMetrics(
+				config.isNetworkDetailedMetrics(),
+				outputMetricGroup,
+				outputMetricGroup.addGroup(METRIC_GROUP_BUFFERS),
```
Might be better to cover all the details in `NettyShuffleMetricFactory`. I mean the logic of `outputMetricGroup.addGroup(METRIC_GROUP_BUFFERS)` here.

azagrebin replied:
I can add overloaded methods for this in `NettyShuffleMetricFactory`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+
+			MetricGroup networkInputGroup = ownerContext.getInputGroup();
+			@SuppressWarnings("deprecation")
+			InputChannelMetrics inputChannelMetrics = new InputChannelMetrics(networkInputGroup, ownerContext.getParentGroup());
```
Maybe we could delay the creation of `InputChannelMetrics` in the `SingleInputGateFactory#createInputChannels`, then the following would be 
`SingleInputGate inputGate = singleInputGateFactory.create(ownerContext, icdd, partitionProducerStateProvider)` instead. 
To do so we could make the current `createInputGates` simple and make all the things done when required.

azagrebin replied:
The gate creation loop will create `InputChannelMetrics` multiple times, I would avoid this because `InputChannelMetrics` are shared by gates at the moment.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java:
```diff
@@ -228,91 +211,111 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
+			registerInputMetrics(
+				config.isNetworkDetailedMetrics(),
+				networkInputGroup,
+				networkInputGroup.addGroup(METRIC_GROUP_BUFFERS),
```
ditto: for the issue of `networkInputGroup.addGroup(METRIC_GROUP_BUFFERS)`

zhijiangW commented:
Thanks for the updates @azagrebin .
I left some final nit comments relevant with new introduced context.

azagrebin commented:
Thanks @zhijiangW , I have addressed comments

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java:
```diff
@@ -0,0 +1,167 @@
+	 * @param parentGroup parent of shuffle specific metric group
+	 * @return context of the shuffle input/output owner used to create partitions or gates belonging to the owner
+	 */
+	ShuffleIOOwnerContext createShuffleIOOwnerContext(
```
it would be useful the partition lifecycle management if this method would also accept the JobID..

azagrebin replied:
Shuffle service does not need it at the moment, would be nice actually to keep it decoupled from the job id, I would suggest we discuss it at the corresponding partition lifecycle PR and add it if needed.

zentol replied:
see this discussion then: https://github.com/apache/flink/pull/8687#discussion_r292370552

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java:
```diff
@@ -0,0 +1,167 @@
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the owner
+	 * @return collection of the {@link ResultPartitionWriter ResultPartitionWriters}
+	 */
+	Collection<P> createResultPartitionWriters(
```
have you considered moving this method and `createInputGates` into the `ShuffleIOOwnerContext`, and only pass this into the task instead of the entire `ShuffleEnvironment`? The task only requires access to these 2 methods, while any other users of the `ShuffleEnvironment` doesn't need these.

azagrebin replied:
This is an interesting idea, we can consider it as a follow-up issue.

azagrebin commented:
Rebased on #8680
